### ISSUE TITLE
Determine and update where to publish `VidDisperseSend`

### ISSUE THREAD

**Issue by shenkeyao** [EspressoSystems/HotShot#1817]
We should determine whether `VidDisperseSend` is supposed to be published in the transaction task as is now, or in the DA task, together with `DAProposalSend`. We may need to update the structure of the `BlockReady` event.

**Comment by ggutoski** [comment]
I'd appreciate an opinion here from @bfish713 when he returns from vacation.

**Comment by ggutoski** [comment]
Yesterday we discussed an idea to create a new task for VID to separate it from the DA and transactions tasks. The idea is: transactions task collect transactions, publish `BlockReady` event. Both DA and VID tasks listen for this event, then do their jobs. All the ugly VID work is done inside the VID task.

Some complications:
- `BlockReady` needs a `BlockPayload`, which is `Committable`, so we need the VID commitment at the beginning (ie. in the transactions task).
- We could eliminate the need for the VID commitment at the beginning if we change `BlockReady` to include only a `Vec<Transaction>`. (Or if we somehow manage to remove `Committable` from `BlockPayload`). Buuuuuuuuut.......
- The DA task also needs the VID commitment because that's what's signed as part of the DA proposal it creates. Soooooo....
- Seems like we're stuck in a situation where we need the VID commitment inside the transactions task. I see no way to avoid this situation.

**Comment by ggutoski** [comment]
# Where to publish the `VidDisperseSend` event?

I suggest we continue to publish `VidDisperseSend` in the transactions task until we get a new task for VID. At that time, it might make sense to move publication of this event to the new VID task. But it's complicated. I explain this complexity below, then I present a new proposal.

Recall discussion from https://github.com/EspressoSystems/HotShot/issues/1817#issuecomment-1741192171 :
- DA task needs the VID commitment
- Thus, VID commitment must be included in `BlockReady` event.
- Thus, VID commitment must be computed *before* `BlockReady` event is published.
- `BlockReady` event is published inside the transactions task, so VID commitment must be computed inside this task.
- VID commitment is expensive, but it comes for free when we compute VID disperse data.
- Thus, we should compute all the VID disperse data inside the transactions task, too.
- Because we compute all the VID disperse data inside the transactions task, the natural thing to do is publish the `VidDisperseSend` event inside the transactions task. 

Given the above, the choice of whether to publish `VidDisperseData` inside the transactions task vs the VID task is largely a cosmetic choice. In all cases, VID disperse data is computed inside the transactions task. If we prefer to publish the `VidDisperseEvent` outside the transactions task then we could (for example) include this data in the `BlockReady` event---then the VID task could read this data from that event and publish `VidDisperseData`. None of this changes the fact that all the VID work is done inside the transactions task---the only question is whether to publish `VidDisperseData` at the moment we compute it inside the transactions task, or to push the data somewhere else so we can instead publish it from the VID task.

# Proposal: new event

Currently the DA task listens for `BlockReady` from the transactions task. Instead, I propose the following flow:
- VID task listens for `BlockReady` from the transactions task.
- VID task computes VID commitment + disperse data, then publishes 2 events: (i) the `VidDisperseSend` event as usual, and (ii) a new `BlockReadyWithVidCommitment` event (but choose a better name). This new event includes (i) block payload data from `BlockReady`, and (ii) VID commitment.
- DA task does *not* listen for `BlockReady`. Instead DA task listens for `BlockReadyWithVidCommitment`, which has all the info it needs.

Advantages:
- All VID-specific computation is done inside the VID task.
- The transactions task reverts to its original minimal purpose.

**Comment by shenkeyao** [comment]
Maybe rename `BlockReady` to `TransactionsReady` since it's building the transactions only, and `BlockReadyWithVidCommitment` could be `BlockReady`.

**Comment by ggutoski** [comment]
This workflow needs to be changed. We recently discussed that the DA committee in fact *does not need* the VID commitment. Why? Because they can't trust it anyway. Instead each DA committee member must re-compute the VID commitment C for itself and then sign an attestation to C.

## New workflow proposed by Jeb

So the workflow can be like: transactions task builds block and encodes to bytes. Then, in parallel:

DA:
* transactions -bytes-> DA leader
* DA leader -bytes-> DA committee
* DA committee computes VID commitment (`commit_only`) and signs it
* DA committee -signature-> DA leader

VID:
* transactions -bytes-> VID leader
* VID leader computes commitment and shares (`disperse`)
* VID leader -disperse-> replicas
* VID leader -commitment-> consensus leader

**Comment by elliedavidson** [comment]
New design: 
* Transactions task - collects transactions and sequences them. Emits a `TransactionsSequenced` event
* VID task - listens for `TransactionsSequenced`, calculates VID upon receiving that event. Emits `BlockReady` event
* VID task - listens for `BlockReady`, then sends out the `VIDProposalSend` event
* DA task - listens for `TransactionsSequenced`, creates DA proposal upon receiving that event. Emits `DAProposalSend`

**Comment by rob-maron** [comment]
Do we still expect the transactions task to calculate the block payload?

I think this makes the most sense as the DA task also requires the block payload and metadata - but just making sure.

**Comment by ggutoski** [comment]
> Do we still expect the transactions task to calculate the block payload?

I say yes. This is dead easy though: just call the trait method `BlockPayload::build`.

**Comment by shenkeyao** [comment]
> > Do we still expect the transactions task to calculate the block payload?
> 
> I say yes. This is dead easy though: just call the trait method `BlockPayload::build`.

+1! Btw it's `BlockPayload::from_transactions` after the renaming.
